util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <config.h>
+
+#include <time.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <assert.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <netdb.h>
+#include <poll.h>
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_eq.h>
+#include <rdma/fi_errno.h>
+
+
+#ifndef FT_FIVERSION
+#define FT_FIVERSION FI_VERSION(1,3)
+#endif
+
+#ifdef __APPLE__
+#include "osx/osd.h"
+#elif defined __FreeBSD__
+#include "freebsd/osd.h"
+#endif
+
+
+enum precision {
+	NANO = 1,
+	MICRO = 1000,
+	MILLI = 1000000,
+};
+
+enum {
+	FT_OPT_ACTIVE		= 1 << 0,
+	FT_OPT_ITER		= 1 << 1,
+	FT_OPT_SIZE		= 1 << 2,
+	FT_OPT_VERIFY_DATA	= 1 << 3,
+};
+
+
+struct ft_opts {
+	char *src_port;
+	char *dst_port;
+	char *src_addr;
+	char *dst_addr;
+	int iterations;
+	int transfer_size;
+	int sizes_enabled;
+	int options;
+	int argc;
+	char **argv;
+};
+
+#define ADDR_OPTS "b:p:s:"
+#define INFO_OPTS "n:f:e:"
+#define CS_OPTS ADDR_OPTS "I:S:"
+#define BENCHMARK_OPTS "v"
+
+#define INIT_OPTS (struct ft_opts) \
+	{	.options = 0, \
+		.iterations = 1000, \
+		.transfer_size = 1024, \
+		.sizes_enabled = FT_DEFAULT_SIZE, \
+		.argc = argc, .argv = argv \
+	}
+
+#define FT_STR_LEN 32
+#define FT_MAX_CTRL_MSG 64
+#define FT_MR_KEY 0xC0DE
+#define FT_MSG_MR_ACCESS (FI_SEND | FI_RECV)
+#define FT_RMA_MR_ACCESS (FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE)
+
+#define FT_PRINTERR(call, retv) \
+	do { fprintf(stderr, call "(): %s:%d, ret=%d (%s)\n", __FILE__, __LINE__, \
+			(int) retv, fi_strerror((int) -retv)); } while (0)
+
+#define FT_LOG(level, fmt, ...) \
+	do { fprintf(stderr, "[%s] %s:%d: " fmt "\n", level, __FILE__, \
+			__LINE__, ##__VA_ARGS__); } while (0)
+
+#define FT_LOG_(level, fmt, ...) \
+	do { fprintf(stderr, "[%s] %s:%d: " fmt, level, __FILE__, \
+			__LINE__, ##__VA_ARGS__); } while (0)
+
+#define FT_ERR(fmt, ...) FT_LOG("error", fmt, ##__VA_ARGS__)
+#define FT_WARN(fmt, ...) FT_LOG("warn", fmt, ##__VA_ARGS__)
+
+int _activate_debug = 0;
```

Comment:
Symbols beginning with a single underscore are in the namespace of the compiler, right?